### ISSUE TITLE
[torch.export] Dim discrete/finite set

### ISSUE THREAD

**Issue by bhack** [pytorch/pytorch#136119]
### 🚀 The feature, motivation and pitch

Is it possible to introduce `dynamic_shape` discrete/finite set?
Suppose that we need to export a model for a set of well known input resolution at inference. 
It seems currently there is not way to do it a we could define only min and max for Dim. How is it going to create optimized code if we only specify the full range? It seems to me too much relaxed

**Comment by avikchaudhuri** [comment]
No, not possible today. Also I don't an easy way of doing this with the underlying symbolic shapes subsystem, because it accumulates constraints in a conjunctive way ("and") and a discrete set is like a disjunction ("or"), which will probably blow up the complexity of reasoning.

Your best alternative is to export with different example inputs, one for each concrete shape in this set, and write a wrapper / dispatcher than calls the corresponding exported program based on the concrete shape of the input.

A possible alternative feature request would be to package up such multiple exported programs into a single artifact (basically upstreaming this kind of wrapper / dispatcher given a set of different example inputs).

cc @ezyang

**Comment by bhack** [comment]
Isn't dynamo defacto doing something like this with jit on runtime? How is it going to manipulate guards at runtime?

**Comment by avikchaudhuri** [comment]
Yes dynamo has a cache that maps guards to compiled code, and something similar is needed here. 

Note that export is (a) technically independent of dynamo, because non-strict doesn't use dynamo for tracing (b) more importantly, it is generating code aot so needs to "inline" this cache into the generated code.

**Comment by bhack** [comment]
Yes something like that and we also need to be able to target specific  HW when we approach with A0T for other optimizations. 
If not we need to export (with auto tune?) only on the specific HW.

Other that form optimizing performances in exported code I think that classical bucketing or padding could have its own overhead so often it is a suboptimal solution.

E.g. https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/torch-neuron/bucketing-app-note.html

**Comment by bhack** [comment]
Yes something like that and we also need to be able to target specific  HW when we approach with A0T for other optimizations. 
If not we need to export (with auto tune?) only on the specific HW.

Other that form optimizing performances in exported code I think that classical bucketing or padding could have its own overhead so often it is a suboptimal solution (E.g. https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/torch-neuron/bucketing-app-note.html)

**Comment by bhack** [comment]
Also, honestly it isn't so much clear what it is then the logic of using different backends at:

https://pytorch.org/tutorials/intermediate/torch_export_tutorial.html#running-the-exported-program


Probably in this topic we need also to better explain and document for the end-user the interplay between guards injected in the export phase and what we could optimize when we load the export program.
Cause it seems that eventually a dynamo backend is useless in that post-loading stage of the tutorial.

**Comment by pianpwk** [comment]
closed duplicate https://github.com/pytorch/pytorch/issues/139307 in favor of this issue
